### DISCUSSION TITLE
Are the contents of `cape-dict-file` cached?

### DISCUSSION THREAD

**Discussion by karthink** [minad/cape#141] [Q&A]
Hi minad,

Just something I'm curious about -- from a cursory read of the `cape-dict` code, it looks like you run a grep command every time you want to complete from `cape-dict-file`.  On the other hand, I know that the dynamic completion table code is quite complex and might be doing internal caching.

So I was curious: `cape-dict` is quite fast but still slows down for me sometimes, so is the word list already cached in memory by the completion system?  If not, do you think it will help to cache it?

-  My `cape-dict-file` is [Peter Norvig's list](https://norvig.com/ngrams/count_1w.txt) of the top 300,000 English words sorted by frequency.
- I have a completion-category-override in place for the `cape-dict` category to do `orderless-literal-prefix` style completion only.

**Comment by minad** [comment]
No, the entire dictionary is not cached. The dynamic completion table only caches the last candidates as long as they remain valid. I am not sure if we can do much better - remember that the OS caches grep and the dictionary file in memory if heavily accessed. Running `all-completions` on all 300k candidates will be too slow.

**Comment by karthink** [comment]
> No, the entire dictionary is not cached. The dynamic completion table only caches the last candidates as long as they remain valid. I am not sure if we can do much better - remember that the OS caches grep and the dictionary file in memory if heavily accessed. Running `all-completions` on all 300k candidates will be too slow.

I didn't know the OS caches grep and the file -- that could be why it's as fast as it currently is.

FWIW, I set cape-dict-limit to 10 and truncated the words file to the first 50,000 words now, and this seems to have helped.  (The words file is sorted in order of frequency of usage.)

**Reply by minad** [reply]
> I didn't know the OS caches grep and the file -- that could be why it's as fast as it currently is.

Well, I mean the OS file system cache. It caches everything which is accessed often.

> FWIW, I set cape-dict-limit to 10 and truncated the words file to the first 50,000 words now, and this seems to have helped.  (The words file is sorted in order of frequency of usage.)

I use the same, but with 300k and 600k files without big issues. However I don't use aggressive auto completion settings.
